packages/rich-text/src/xpath.js

```diff
@@ -0,0 +1,111 @@
+function append( parent, object ) {
+	object.parent = parent;
+	parent.children = parent.children || [];
+	parent.children.push( object );
+	return object;
+}
+
+function getLastChild( { children } ) {
+	return children && children[ children.length - 1 ];
+}
+
+const matchNodePath = /([a-z0-9()]+)\[(\d)+]/;
+
+function isTextNode( child ) {
+	return child.hasOwnProperty( 'text' );
+}
+
+function isNodeType( type, child ) {
+	return child.hasOwnProperty( 'type' ) && type === child.type;
+}
+
+function findNodeWithIndex( type, index, children ) {
+	let count = 0;
+	let foundNode = false;
+
+	children.forEach( ( child ) => {
+		if (
+			( type === 'text()' && isTextNode( child ) ) ||
+			isNodeType( type, child )
+		) {
+			if ( count === index ) {
+				foundNode = child;
+			}
+
+			count++;
+		}
+	} );
+
+	return foundNode;
+}
+
+/**
+ * Matches an XPath to a certain value.
+ *
+ * Has certain assumptions about the XPath:
+ *
+ * * The XPath is continuous: So all nodes until the deepest node are included.
+ * * The XPath starts from the root element of the RichText.
+ * * The leaf node should always be a text() node.
+ *
+ * @param {Object} record Rich text value.
+ * @param {string} xpath XPath to match.
+ *
+ * @return {number} The position of the matched 'element'.
+ */
+export function matchXPath( record, xpath ) {
+	const { formats, text } = record;
+
+	const formatsLength = formats.length + 1;
+	const tree = {};
+
+	append( tree, { text: '', pos: 0 } );
+
+	for ( let i = 0; i < formatsLength; i++ ) {
```

Comment:
Curious if you could use `toTree` here.

Reply:
I've rewritten it to use `toTree`. I had to add positional information to the callbacks to be able to do it.